### PR TITLE
Enhancements proposal

### PR DESCRIPTION
**Pull request by molaux**
I share you my personal modifications to this project, in case it can be useful to others than me.

It should be backward compatible, since default configuration options doesn't change meaningfully the generated files.

### CODE REVIEW

**Comment by molaux** [comment]
I have written a [complete example](https://github.com/molaux/graphql-sequelize-r-example) for testing purpose. It could be added to the documentation.

**Comment by tohenk** [comment]
Hi @molaux , can you rebase this PR?
The last commit allow CI to check for errors, also there is some new feature already added such as common table property.

**Comment by molaux** [comment]
Hi @tohenk, 
I rebased my work and updated my [test repo](https://github.com/molaux/sequelize-graphql-schema-builder-example).

File: README.md [hunk 1]
```diff
@@ -26,10 +26,51 @@ You then can invoke the CLI script using `vendor/bin/mysql-workbench-schema-expo
...
+    ```
+    Default is `false`.
+
+  * `generateForeignKeysFields`
+
+    Whether or not to generate foreign keys fields and indexes. You could want to delegate it to association method (one could experiment relations creation order problems when it is present).
+
+    Default is `true`.
+
+  * `useTimestamps`
```
Comment (tohenk) [inline README.md:70]: This property for table can be dropped, as it now can be supplied using dc74287494fd0788e369ad802cffbedeba42c560.

File: lib/Sequelize6/Formatter.php [hunk 2]
```diff
@@ -33,6 +33,26 @@
 
 class Formatter extends BaseFormatter
 {
+    const CFG_USE_SEMICOLONS                        = 'useSemicolons';
+    const CFG_GENERATE_ASSOCIATION_METHOD           = 'generateAssociationMethod';
+    const CFG_GENERATE_FOREIGN_KEYS_FIELDS          = 'generateForeignKeysFields';
+    const CFG_USE_TIMESTAMPS                        = 'useTimestamps';
```
Comment (tohenk) [inline lib/Sequelize6/Formatter.php:40]: See above.

File: lib/Sequelize6/Formatter.php [hunk 3]
```diff
@@ -33,6 +33,26 @@
...
+    /**
+     * (non-PHPdoc)
+     * @see \MwbExporter\Formatter\Formatter::init()
+     */
+    protected function init()
+    {
+        parent::init();
+        $this->addConfigurations(array(
+            static::CFG_INDENTATION                   => 4,
```
Comment (tohenk) [inline lib/Sequelize6/Formatter.php:51]: It has already been registered in parent initialization, https://github.com/mysql-workbench-schema-exporter/node-exporter/blob/master/lib/Formatter.php#L44.

File: lib/Sequelize6/Model/Table.php [hunk 4]
```diff
@@ -91,30 +94,50 @@ protected function writeBody(WriterInterface $writer)
...
         $result = array_merge([
-            'sequelize' => $this->getJSObject('sequelize', false, true),
-            'modelName' => $this->getModelName(),
-            'tableName' => $this->getRawTableName(),
-            'indexes' => count($indexes = $this->getIndexes()) ? $indexes : null,
+            'sequelize'         => $this->getJSObject('sequelize', false, true),
+            'modelName'         => $this->getModelName(),
+            'tableName'         => $this->getRawTableName(),
+            'indexes'           => count($indexes = $this->getIndexes()) ? $indexes : null,
+            'timestamps'        => $this->getConfig()->get(Formatter::CFG_USE_TIMESTAMPS),
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:137]: timestamps, underscored, and synOnAssociation can be dropped, see above.

File: lib/Sequelize6/Model/Table.php [hunk 5]
```diff
@@ -180,15 +215,176 @@ protected function getIndexes()
...
+                    : $local->getOwningTable()->getModelName()), $relatedAlias));
+            } else {
+                $as = $this->getNaming($local->isManyToOne() ? $this->pluralize($local->getOwningTable()->getModelName()) : $local->getOwningTable()->getModelName());
+            }
+
+            if ($as === "" || $as === $local->getOwningTable()->getModelName()) {
+                $as = null;
+            }
+
+            $options = array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:337]: Use short array syntax.

File: lib/Sequelize6/Model/Table.php [hunk 6]
```diff
@@ -180,15 +215,176 @@ protected function getIndexes()
...
+                'foreignKey'    => array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:338]: This one too.

File: lib/Sequelize6/Model/Table.php [hunk 7]
```diff
@@ -180,15 +215,176 @@ protected function getIndexes()
...
+            $as = null;
+            $foreignColumnName = $foreign->getForeign()->getColumnName();
+            if ($relatedColumnName) {
+                $relatedAlias = preg_replace("/_?${foreignColumnName}_?/", '', $relatedColumnName);
+                $as = $this->getNaming($relatedAlias);
+                if ($as === $foreign->getReferencedTable()->getModelName()) {
+                    $as = null;
+                }
+            }
+            $options = array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:415]: This one.

File: lib/Sequelize6/Model/Table.php [hunk 8]
```diff
@@ -180,15 +215,176 @@ protected function getIndexes()
...
+                'foreignKey'    => array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:416]: This one.

File: lib/Sequelize6/Model/Table.php [hunk 9]
```diff
@@ -180,15 +215,176 @@ protected function getIndexes()
...
+                    $foreign->getReferencedTable()->getModelName(),
+                    $this->getJSObject($options));
+            $firstAssociation = false;
+        }
+
+        // N <=> M associations
+        foreach ($this->getTableM2MRelations() as $relation) {
+            $this->getDocument()->addLog(sprintf('  Writing setter/getter for N <=> N "%s"', $relation['refTable']->getModelName()));
+
+            $options = array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:458]: This one.

File: lib/Sequelize6/Model/Table.php [hunk 10]
```diff
@@ -180,15 +215,176 @@ protected function getIndexes()
...
+                'through'       => $relation['reference']->getOwningTable()->getRawTableName(),
+                'foreignKey'    => array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:460]: This one.

File: lib/Sequelize6/Formatter.php [hunk 11]
```diff
@@ -33,6 +33,26 @@
...
+    const CFG_GENERATE_FOREIGN_KEYS_FIELDS          = 'generateForeignKeysFields';
+    const CFG_USE_TIMESTAMPS                        = 'useTimestamps';
+    /**
+     * (non-PHPdoc)
+     * @see \MwbExporter\Formatter\Formatter::init()
+     */
+    protected function init()
+    {
+        parent::init();
+        $this->addConfigurations(array(
```
Comment (tohenk) [inline lib/Sequelize6/Formatter.php:50]: Use short array syntax.

File: lib/Sequelize6/Model/Table.php [hunk 12]
```diff
@@ -269,7 +301,7 @@ protected function writeAssociations(WriterInterface $writer)
                 $as = $this->getNaming($local->getOwningTable()->getModelName(), null, true);
             }
 
-            if ($as === "" || $as === $local->getOwningTable()->getModelName()) {
+            if ($as === "" || $as === $local->getOwningTable()->getModelName() || $relatedAlias === $foreignColumnName) {
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php]: Change to single quote.

File: lib/Sequelize6/Model/Table.php [hunk 13]
```diff
@@ -243,7 +243,7 @@ protected function getIndexes()
     }
 
     protected function getConstraints() {
-        $constraints;
+        $constraints = array();
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:252]: Use short array syntax.

File: lib/Sequelize6/Model/Table.php [hunk 14]
```diff
@@ -444,6 +444,23 @@ protected function writeAssociations(WriterInterface $writer)
...
+    /**
+     * Inject a many to many relation into referenced table.
+     *
+     * @param \MwbExporter\Model\ForeignKey $fk1
+     * @param \MwbExporter\Model\ForeignKey $fk2
+     * @return \MwbExporter\Model\Table
+     */
+    protected function injectManyToMany(ForeignKey $fk1, ForeignKey $fk2)
+    {
+        $fk1->getReferencedTable()->setManyToManyRelation(array(
```
Comment (tohenk) [inline lib/Sequelize6/Model/Table.php:491]: Use short array sytax.

File: lib/Formatter.php [hunk 15]
```diff
@@ -74,7 +74,7 @@ public function getTableProp()
     {
         $prop = [
             'timestamps' => false,
-            'underscored' => false,
+            'underscored' => true,
```
Comment (tohenk) [inline lib/Formatter.php:77]: It's better to leave underscored as false by default.

**Pull request merged**